report/topology.go

```diff
@@ -98,83 +71,28 @@ func (n NodeMetadatas) Merge(other NodeMetadatas) NodeMetadatas {
-	EgressPacketCount  *uint64 `json:"egress_packet_count,omitempty"`
-	IngressPacketCount *uint64 `json:"ingress_packet_count,omitempty"`
-	EgressByteCount    *uint64 `json:"egress_byte_count,omitempty"`  // Transport layer
-	IngressByteCount   *uint64 `json:"ingress_byte_count,omitempty"` // Transport layer
-	MaxConnCountTCP    *uint64 `json:"max_conn_count_tcp,omitempty"`
-}
-
-// Copy returns a value copy of the EdgeMetadata.
-func (e EdgeMetadata) Copy() EdgeMetadata {
-	return EdgeMetadata{
-		EgressPacketCount:  cpu64ptr(e.EgressPacketCount),
-		IngressPacketCount: cpu64ptr(e.IngressPacketCount),
-		EgressByteCount:    cpu64ptr(e.EgressByteCount),
-		IngressByteCount:   cpu64ptr(e.IngressByteCount),
-		MaxConnCountTCP:    cpu64ptr(e.MaxConnCountTCP),
-	}
-}
-
-func cpu64ptr(u *uint64) *uint64 {
-	if u == nil {
-		return nil
-	}
-	value := *u   // oh man
-	return &value // this sucks
-}
-
-// Merge merges another EdgeMetadata into the receiver and returns the result.
-// The receiver is not modified. The two edge metadatas should represent the
-// same edge on different times.
-func (e EdgeMetadata) Merge(other EdgeMetadata) EdgeMetadata {
-	cp := e.Copy()
-	cp.EgressPacketCount = merge(cp.EgressPacketCount, other.EgressPacketCount, sum)
-	cp.IngressPacketCount = merge(cp.IngressPacketCount, other.IngressPacketCount, sum)
-	cp.EgressByteCount = merge(cp.EgressByteCount, other.EgressByteCount, sum)
-	cp.IngressByteCount = merge(cp.IngressByteCount, other.IngressByteCount, sum)
-	cp.MaxConnCountTCP = merge(cp.MaxConnCountTCP, other.MaxConnCountTCP, max)
-	return cp
-}
-
-// Flatten sums two EdgeMetadatas and returns the result. The receiver is not
-// modified. The two edge metadata windows should be the same duration; they
-// should represent different edges at the same time.
-func (e EdgeMetadata) Flatten(other EdgeMetadata) EdgeMetadata {
-	cp := e.Copy()
-	cp.EgressPacketCount = merge(cp.EgressPacketCount, other.EgressPacketCount, sum)
-	cp.IngressPacketCount = merge(cp.IngressPacketCount, other.IngressPacketCount, sum)
-	cp.EgressByteCount = merge(cp.EgressByteCount, other.EgressByteCount, sum)
-	cp.IngressByteCount = merge(cp.IngressByteCount, other.IngressByteCount, sum)
-	// Note that summing of two maximums doesn't always give us the true
-	// maximum. But it's a best effort.
-	cp.MaxConnCountTCP = merge(cp.MaxConnCountTCP, other.MaxConnCountTCP, sum)
-	return cp
-}
-
 // NodeMetadata describes a superset of the metadata that probes can collect
 // about a given node in a given topology.
 type NodeMetadata struct {
 	Metadata  map[string]string
 	Counters  map[string]int
 	Adjacency IDList
+	EdgeMetadatas
 }
 
 // MakeNodeMetadata creates a new NodeMetadata with no initial metadata.
 func MakeNodeMetadata() NodeMetadata {
-	return MakeNodeMetadataWith(map[string]string{})
+	return NodeMetadata{
+		Metadata:      map[string]string{},
+		Counters:      map[string]int{},
+		Adjacency:     MakeIDList(),
+		EdgeMetadatas: EdgeMetadatas{},
```

Comment:
So we're not going to make IDList not have a merge and a copy function, therefore this would imply a type for Counters and Metadata, which I'm fine with, but when I introduced Metadata you weren't keen on aliasing map[string]string...